torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1823,6 +1823,15 @@ def call_compiled_backward():
                     )
                     return tuple(out)
 
+                # Backward with forward inputs mutations is not supported in double backward.
+                if torch.is_grad_enabled() and any(
+                    i.joint_mutates_data and not i.mutates_data
```

Comment:
hmm two things about this assertion:

(1) if we have a graph that mutates a tensor that does **not** require grad during the backward, this assert will fail incorrectly (even if `is_grad_enabled() == True` during the backward and we see an input mutation, as long as that input does not require grad it is ok to keep it in the graph)

(2) this assert will never fire if we have a tensor that is mutated in both the forward **and** the backward. but that seems wrong.

Reply:
About (2): Potentially in future we can retrace backward_module after partitioning and by version counter determine if bwd mutated tensor that was mutated in forward.